[flutter_tools] Support zipped application bundles for macOS

rokob opened the pull request:
## Description

It is not possible to directly produce a directory (.app) in some build systems
but rather it must be zip'ed before being passed to the tool for
running. This adds support for attempting to extract an application
bundle from a zip file if the bundle is not already a directory. This
uses very similar code from lib/src/application_package.dart which is
used for extracting an ipa for iOS.

## Tests

I added the following tests:

macos/application_package.dart had no tests so I added one which is modeled after the application_package.dart test for iOS. It tests this behavior that I am adding and some of the other behavior around it. But it is not a comprehensive test of the entire file.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

rokob committed:
[flutter_tools] Support zipped application bundles for macOS

It is not possible to directly produce a directory (.app) in some build systems
but rather it must be zip'ed before being passed to the tool for
running. This adds support for attempting to extract an application
bundle from a zip file if the bundle is not already a directory. This
uses very similar code from lib/src/application_package.dart which is
used for extracting an ipa for iOS.

mehmetf commented:
I wonder if this is better done via a customizable callback rather than implemented in tools directly. Build systems might produce various artifacts (zips, tars, rars, what have you) and it is unreasonable to expect flutter_tools to handle all of these. We can then implement this callback in g3 as part of our custom tooling.

rokob commented:
The closest point of customization that currently exists would be to replace the ApplicationPackageFactory (which we already do in g3) and then override for platform darwin_x64. That could be made to work without having to change anything here, but that felt like it needs quite a bit of duplication and this seemed pretty minimal and actually inline with what we already do for iOS.

Otherwise, we could add something to the AppContext specifically for getting the executable, id, and bundle for macos and have the default just be the current implementation of _executableAndId and otherwise override it. This would amount to about the same thing as doing the ApplicationPackageFactory override.

The other option is to have a callback with signature `Directory? Function(FileSystemEntity)` which we call if the FileSystemEntity (applicationBundle) is not a directory. That would be the most targeted hook albeit very specific.

mehmetf commented:
I leave it to @jonahwilliams . The fact that the override needs to happen in a static context doesn't make things easy. I was mostly thinking of the last option: Derive the bundle directory given the file system entity. You can make it more generic by not making it a fallback option. Just 

```dart
Directory appBundle = globals.bundleProcessor.getAppBundle(FileSystemEntity);
```

which defaults to the current implementation of just returning a cast. We can do something else in g3.

If there's an option of not doing this in a static context, we can expose this as an overridable function instead.

mehmetf commented:
Flutter follows a different formatter than what g3 uses. Please revert the lines that were changed by the formatter.

rokob committed:
Introduce BundleProcessor with default behavior

This introduces the BundleProcessor class which by default just checks
that the input bundle is a directory and does the type cast. This is the
hook I need to be able to handle non-directory applicatin bundles
(namely zip files). I added a test for overriding this with the
unzipping behavior to ensure that this indeed is usable for what I need.

rokob commented:
Ok I reverted the format changes. I added a class BundleProcessor which captures the current behavior and is overridable in the context. The tests that I added for the normal case still pass with the default BundleProcessor. My additional test for the unzipping behavior also pass when I override the BundleProcessor with one that does the unzipping like I original add inline.

The tests I added also found a bug in the current implementation because `_executableFromBundle` returns null upon failure, but fromPrebuiltApp unconditionally reads properties (id and executable) from the result of that function. I added a null check to fix that.

mehmetf commented:
I like it :) However, as I said, the the final verdict lies with @jonahwilliams

jonahwilliams reviewed (commented):
While we don't have an immediate use case for supporting a zipped macOS application right now, it seems like a perfectly reasonable thing to support.

Introducing another injected interface is not the worst thing in the world, the BundleProcessor is essentially doing the same thing that the macOSApp class was doing before - but now it is split up into separate classes. I think this will be harder to maintain going forward.

rokob committed:
Remove BundleProcessor and cleanup code and tests

This reverts the prior commit where I added BundleProcessor to keep the
code inline and just adds support for unzipping the bundle inline. I
changed the name of _ExecutableAndIdAndBundle to _BundleInfo because the
other name seemed ridiculous. I also added another test for one of the
failure modes.

rokob commented:
I went ahead and went back to the form of doing the unzipping in case the bundle is not a directory and got rid of the BundleProcessor because I agree that was spreading code out a bit more than necessary if you are willing to have the unzipping as part of this process.

jonahwilliams commented on packages/flutter_tools/lib/src/macos/application_package.dart:
```diff
@@ -58,8 +63,26 @@ abstract class MacOSApp extends ApplicationPackage {
-      return null;
+      // Try to unpack as a zip.
+      final Directory tempDir = globals.fs.systemTempDirectory.createTempSync('flutter_app.');
+      shutdownHooks.addShutdownHook(() async {
```
I don't believe this is necessary because the tool should delete an outstanding directories during shutdown. (unless it receives a sigkill, but then the shutdown hook won't run anyway)

rokob replied:
I see that now in systemTempDirectory, cool.

jonahwilliams commented on packages/flutter_tools/lib/src/macos/application_package.dart:
```diff
@@ -58,8 +63,26 @@ abstract class MacOSApp extends ApplicationPackage {
+        return null;
+      }
+      try {
+        bundleDir = tempDir
```
rather than catch the state error, I would use the firstWhere orElse:() => null pattern

rokob replied:
firstWhere handles the case of no application bundle but it does not handle the case of multiple bundles in the zip file. To handle both without the StateError we could check that the length of the iterable returned by where(_isBundleDirectory) is equal to one. But using singleWhere we actually get a StateError thrown for both of those cases and it seems like this actually looks a bit nicer.

jonahwilliams replied:
Yeah, that is a good point. SGTM

jonahwilliams reviewed (approved):
LGTM with two nits.

Sorry this took so long to review!

rokob committed:
remove shutdownHook because temp dirs are already cleaned up automatically
